### PR TITLE
[Filebeat] Refactor mqtt input

### PR DESCRIPTION
**Pull request by mtojek**
Issue: https://github.com/elastic/beats/issues/15982

This PR refactor recently developed mqtt input (https://github.com/elastic/beats/pull/15287) to provide the following features:
* automatic reconnect on connection lost/failure
* TLS support with `tlscommon.TLS`
* improved shutdown procedure (first disconnect, then wait for `onMessage` handler functions to finish processing
* supported and enabled logging in the library `github.com/eclipse/paho.mqtt.golang`

I removed data decoding to MsgPack/JSON as (to my understanding) processors should be responsible for that (see: https://www.elastic.co/guide/en/beats/filebeat/current/decode-json-fields.html)

Filebeat.yml to test this change:
```
filebeat.inputs:
- type: mqtt
  hosts:
    - ssl://mqtt.eclipse.org:8883
  topics:
    - motion

output.elasticsearch:
  hosts: ["localhost:9200"]
```

Sample output with input logging enabled:

```
2020-02-04T12:27:18.647+0100    INFO    [beat]  instance/beat.go:991    Process info    {"system_info": {"process": {"cwd": "/Users/marcin.tojek/go/src/github.com/elastic/beats/filebeat", "exe": "./filebeat", "name": "filebeat", "pid": 71825, "ppid": 25569, "start_time": "2020-02-04T12:27:18.612+0100"}}}
2020-02-04T12:27:18.647+0100    INFO    instance/beat.go:297    Setup Beat: filebeat; Version: 8.0.0
2020-02-04T12:27:18.647+0100    INFO    [index-management]      idxmgmt/std.go:182      Set output.elasticsearch.index to 'filebeat-8.0.0' as ILM is enabled.
2020-02-04T12:27:18.647+0100    INFO    elasticsearch/client.go:174     Elasticsearch url: http://localhost:9200
2020-02-04T12:27:18.648+0100    INFO    [publisher]     pipeline/module.go:110  Beat name: MacBook-Elastic.local
2020-02-04T12:27:18.648+0100    INFO    [monitoring]    log/log.go:118  Starting metrics logging every 30s
2020-02-04T12:27:18.648+0100    INFO    instance/beat.go:438    filebeat start running.
2020-02-04T12:27:18.648+0100    INFO    registrar/registrar.go:145      Loading registrar data from /Users/marcin.tojek/go/src/github.com/elastic/beats/filebeat/data/registry/filebeat/data.json
2020-02-04T12:27:18.649+0100    INFO    registrar/registrar.go:152      States Loaded from registrar: 0
2020-02-04T12:27:18.649+0100    INFO    crawler/crawler.go:72   Loading Inputs: 1
2020-02-04T12:27:18.649+0100    INFO    input/input.go:114      Starting input of type: mqtt; ID: 1514614463779256881
2020-02-04T12:27:18.649+0100    INFO    crawler/crawler.go:106  Loading and starting Inputs completed. Enabled inputs: 1
2020-02-04T12:27:18.649+0100    DEBUG   [mqtt input]    mqtt/input.go:169       Run the input once.     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:19.209+0100    WARN    [libmqtt]       paho.mqtt.golang/memstore.go:137        [store]   memorystore wiped
2020-02-04T12:27:19.209+0100    DEBUG   [mqtt input]    mqtt/input.go:145       Try subscribe to topics: motion {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:19.476+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 176     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:19.635+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 176     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:19.741+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 177     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:19.847+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 180     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:19.848+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 179     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:19.955+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 177     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:20.111+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 177     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:20.221+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 177     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:20.327+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 179     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:20.327+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 177     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:20.437+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 179     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:27:20.480+0100    INFO    pipeline/output.go:95   Connecting to backoff(elasticsearch(http://localhost:9200))
2020-02-04T12:27:20.587+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 177     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
```

Sample output with library logging enabled:
```
2020-02-04T12:26:11.829+0100    INFO    crawler/crawler.go:72   Loading Inputs: 1
2020-02-04T12:26:11.830+0100    INFO    input/input.go:114      Starting input of type: mqtt; ID: 1514614463779256881
2020-02-04T12:26:11.830+0100    INFO    crawler/crawler.go:106  Loading and starting Inputs completed. Enabled inputs: 1
2020-02-04T12:26:11.830+0100    DEBUG   [mqtt input]    mqtt/input.go:169       Run the input once.     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:26:11.830+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:214  [client]  Connect()
2020-02-04T12:26:11.830+0100    DEBUG   [libmqtt]       paho.mqtt.golang/memstore.go:48 [store]   memorystore initialized
2020-02-04T12:26:11.830+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:257  [client]  about to write new connect msg
2020-02-04T12:26:12.287+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:263  [client]  socket connected to broker
2020-02-04T12:26:12.287+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:278  [client]  Using MQTT 3.1.1 protocol
2020-02-04T12:26:12.287+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:485  [net]     connect started
2020-02-04T12:26:12.395+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:503  [net]     received connack
2020-02-04T12:26:12.395+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:347  [client]  client is connected
2020-02-04T12:26:12.395+0100    WARN    [libmqtt]       paho.mqtt.golang/memstore.go:137        [store]   memorystore wiped
2020-02-04T12:26:12.395+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:366  [client]  exit startClient
2020-02-04T12:26:12.395+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:210     [net]     logic started
2020-02-04T12:26:12.395+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:110     [net]     incoming started
2020-02-04T12:26:12.396+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:148     [net]     outgoing started
2020-02-04T12:26:12.395+0100    DEBUG   [mqtt input]    mqtt/input.go:145       Try subscribe to topics: motion {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:26:12.395+0100    DEBUG   [libmqtt]       paho.mqtt.golang/ping.go:27     [pinger]  keepalive starting
2020-02-04T12:26:12.396+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:151     [net]     outgoing waiting for an outbound message
2020-02-04T12:26:12.396+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:213     [net]     logic waiting for msg on ibound
2020-02-04T12:26:12.396+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:740  [client]  enter SubscribeMultiple
2020-02-04T12:26:12.396+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:782  [client]  sending subscribe message, topics:[motion]
2020-02-04T12:26:12.396+0100    DEBUG   [libmqtt]       paho.mqtt.golang/client.go:793  [client]  exit SubscribeMultiple
2020-02-04T12:26:12.396+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:181     [net]     obound priority msg to write, type*packets.SubscribePacket
2020-02-04T12:26:12.396+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:151     [net]     outgoing waiting for an outbound message
2020-02-04T12:26:12.507+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:116     [net]     Received Message
2020-02-04T12:26:12.507+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:217     [net]     logic got msg on ibound
2020-02-04T12:26:12.507+0100    DEBUG   [libmqtt]       paho.mqtt.golang/memstore.go:113        [store]   memorystore del: message1was deleted
2020-02-04T12:26:12.507+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:224     [net]     received suback, id:1
2020-02-04T12:26:12.507+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:228     [net]     granted qoss[0]
2020-02-04T12:26:12.507+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:213     [net]     logic waiting for msg on ibound
2020-02-04T12:26:12.620+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:116     [net]     Received Message
2020-02-04T12:26:12.620+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:217     [net]     logic got msg on ibound
2020-02-04T12:26:12.620+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:240     [net]     received publish, msgId:0
2020-02-04T12:26:12.620+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:241     [net]     putting msg on onPubChan
2020-02-04T12:26:12.620+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:254     [net]     done putting msg on incomingPubChan
2020-02-04T12:26:12.620+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:213     [net]     logic waiting for msg on ibound
2020-02-04T12:26:12.620+0100    DEBUG   [mqtt input]    mqtt/input.go:108       Received message on topic 'motion', messageID: 0, size: 178     {"hosts": ["ssl://mqtt.eclipse.org:8883"]}
2020-02-04T12:26:12.784+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:116     [net]     Received Message
2020-02-04T12:26:12.784+0100    DEBUG   [libmqtt]       paho.mqtt.golang/net.go:217     [net]     logic got msg on ibound
```

### CODE REVIEW

**Comment by mtojek** [comment]
Thank you for the review @faec ! I will fix these issues tomorrow. Also, wondering if this need additional system testing, WDYT?

**Comment by faec** [comment]
> Thank you for the review @faec ! I will fix these issues tomorrow. Also, wondering if this need additional system testing, WDYT?

Yes, there should be some kind of system or integration test before the merge with master, though there's no need to delay this checkin for it since this is just the feature branch.

File: filebeat/input/mqtt/config.go [hunk 1]
```diff
@@ -19,57 +19,34 @@ package mqtt
...
-	DecodePayload  bool          `config:"decode_payload"`
-	SSL            bool          `config:"ssl"`
-	CA             string        `config:"CA"`
-	ClientCert     string        `config:"clientCert"`
-	ClientKey      string        `config:"clientKey"`
-	ClientID       string        `config:"clientID"`
-	WaitClose      time.Duration `config:"wait_close" validate:"min=0"`
-	ConnectBackoff time.Duration `config:"connect_backoff" validate:"min=0"`
+	Hosts  []string `config:"hosts" validate:"required,min=1"`
+	Topics []string `config:"topics" validate:"nonzero,min=1"`
```
[the remark this answers is not in the document]
Comment (mtojek) [reply filebeat/input/mqtt/config.go]: Fixed.

File: filebeat/input/mqtt/input.go [hunk 2]
```diff
@@ -18,45 +18,56 @@
...
 	"github.com/elastic/beats/filebeat/input"
 	"github.com/elastic/beats/libbeat/beat"
 	"github.com/elastic/beats/libbeat/common"
 	"github.com/elastic/beats/libbeat/common/backoff"
 	"github.com/elastic/beats/libbeat/logp"
+)
 
-	"github.com/pkg/errors"
+const (
+	disconnectTimeout = 3 * 1000 // 3000 ms = 3 sec
```
[the remark this answers is not in the document]
Comment (mtojek) [reply filebeat/input/mqtt/input.go]: Fixed.

File: filebeat/input/mqtt/input.go [hunk 3]
```diff
@@ -66,65 +77,109 @@ func NewInput(
...
 }
 
-// Wait should shut down the input and wait for it to complete
-// The disconnect of the client will do this for us
-func (input *mqttInput) Wait() {
-	input.Stop()
+// Wait method waits until event processing is finished.
+func (mi *mqttInput) Wait() {
+	mi.logger.Debug("Wait for the input to finish processing.")
+	mi.inflightMessages.Wait()
```
[the remark this answers is not in the document]
Comment (mtojek) [reply filebeat/input/mqtt/input.go]: Fixed.

**Pull request merged**